### PR TITLE
SNOW-767679: Implement group_by

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/relational_grouped_dataframe.py [hunk 1]
```diff
@@ -189,12 +193,20 @@ def _to_df(self, agg_exprs: List[Expression]) -> DataFrame:
...
+                    ),
+                    analyzer=self._df._session._analyzer,
+                )
+            else:
+                group_plan = SelectStatement(
+                    from_=SelectSnowflakePlan(
+                        snowflake_plan=group_plan, analyzer=self._df._session._analyzer
+                    ),
+                    analyzer=self._df._session._analyzer,
+                )
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/relational_grouped_dataframe.py:209]: Is it possible we don't change this part and do convert a`SelectSnowflakePlan` to `MockSelectSnowflakePlan` in the mock code?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/relational_grouped_dataframe.py]: I moved the logic to `mock_analyzer`, I think that's a good place to intercept and re-wrap the plan into a mock one

File: src/snowflake/snowpark/mock/mock_analyzer.py [hunk 2]
```diff
@@ -428,17 +423,7 @@ def table_function_expression_extractor(
...
-            if isinstance(expr.child, Attribute):
-                self.generated_alias_maps[expr.child.expr_id] = quoted_name
-                for k, v in self.alias_maps_to_use.items():
-                    if v == expr.child.name:
-                        self.generated_alias_maps[k] = quoted_name
-            return alias_expression(
-                self.analyze(expr.child, parse_local_name), quoted_name
-            )
-        if isinstance(expr, UnresolvedAlias):
+        if isinstance(expr, (Alias, UnresolvedAlias)):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/mock_analyzer.py:426]: Is this change related?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/mock_analyzer.py]: bring back the alias logic

**Review by sfc-gh-yixie — APPROVED** [review]
